File: leetcode/day1/findBall.js
// https://leetcode.com/problems/where-will-the-ball-fall/?envType=study-plan&id=level-2
const findBall = (grid) => {
  const isStuck = (row, col) => {
    // outside of grid bounds, so the ball fell through
    if (row < 0 || row > grid.length - 1) return col
    if (col < 0 || col > grid[row].length - 1) return col

    // check if adjacent values are the same
    const cv = grid[row][col]
    const nv = grid[row][col + cv]
    if (cv !== nv) return -1

    // move left or right diagnoal based on cv
    // if cv is 1, it moves right, down
    // if cv is -1, it moves left, down
    return isStuck(row + 1, col + cv)
  }

  const output = []
  for (let i = 0; i < grid[0].length; i++) {
    // check for every column in row 0
    output.push(isStuck(0, i))
  }

  return output
}

describe("FindBall", () => {
  it("returns Array of outcomes", () => {
    let grid = [
      [1, 1, 1, -1, -1],
      [1, 1, 1, -1, -1],
      [-1, -1, -1, 1, 1],
      [1, 1, 1, 1, -1],
      [-1, -1, -1, -1, -1]
    ]
    expect(findBall(grid)).toEqual([1, -1, -1, -1, -1])

    grid = [[-1]]
    expect(findBall(grid)).toEqual([-1])

    grid = [
      [1, 1, 1, 1, 1, 1],
      [-1, -1, -1, -1, -1, -1],
      [1, 1, 1, 1, 1, 1],
      [-1, -1, -1, -1, -1, -1]
    ]
    expect(findBall(grid)).toEqual([0, 1, 2, 3, 4, -1])
  })
})

File: python/find_ball.py
# https://leetcode.com/problems/where-will-the-ball-fall/?envType=study-plan&id=level-2
def find_ball(grid: list[list[int]]) -> list[int]:
    def will_it_fall(grid_i: int , ball_i: int) -> int:
        # ball as fallen through
        if grid_i >= len(grid):
            return 1

        # out of bounds
        if ball_i < 0 or ball_i >= len(grid[0]):
            return -1

        cur_val = grid[grid_i][ball_i]
        next_i = ball_i + cur_val

        # at the edges or at v place
        if next_i < 0 or next_i >= len(grid[0]) or cur_val != grid[grid_i][next_i]:
            return -1

        return will_it_fall(grid_i+1, ball_i + cur_val)

    if len(grid) == 0:
        return []

    return [will_it_fall(0, i) for i in range(len(grid[0]))]

grid = [
    [1, 1, 1, -1, -1],
    [1, 1, 1, -1, -1],
    [-1, -1, -1, 1, 1],
    [1, 1, 1, 1, -1],
    [-1, -1, -1, -1, -1]
]
assert(find_ball(grid) == [1, -1, -1, -1, -1])
